app/lib/permalink_fu.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module PermalinkFu
+  extend ActiveSupport::Concern
+
+  included do
+    scope :by_permalink, ->(permalink) { where(permalink: permalink) }
+  end
+
+  module ClassMethods
+    attr_reader :permalink_options
+    attr_reader :permalink_attributes
+
+    def permalink(attr_names = [], options = {})
+      @permalink_attributes = Array(attr_names)
+      @permalink_options    = options
+      before_validation :create_unique_permalink
+      validates_with PermalinkValidator
+    end
+  end
+
+  private
+
+  def create_unique_permalink
+    base_permalink = permalink.presence || build_permalink_from_attributes
+    self.permalink = generate_unique_permalink(base_permalink)
+  end
+
+  def generate_unique_permalink(base_permalink)
+    # TODO: CONCEPTS THAT I MUST CHECK:
+    # 1: If the permalink must be unique, why doing count of the exact string? it must be always 0 or 1 then...
```

Comment:
And why there is not a unique validation neither in the model nor in the schema?